blockchain/utxocache.go

```diff
@@ -0,0 +1,909 @@
+	b.utxoCache.lastFlushHash = state.lastFlushHash
+	b.utxoCache.lastEvictionHeight = state.lastFlushHeight
+
+	// If state is already caught up to the tip, return as there is nothing to do.
+	if state.lastFlushHash == tip.hash {
+		return nil
+	}
+
+	// Find the fork point between the current tip and the last flushed block.
+	lastFlushedNode := b.index.LookupNode(&state.lastFlushHash)
+	if lastFlushedNode == nil {
+		// panic if the last flushed block node does not exist.  This should never
+		// happen unless the database is corrupted.
+		panicf("last flushed block node hash %v (height %v) does not exist",
+			state.lastFlushHash, state.lastFlushHeight)
+	}
+	fork := b.bestChain.FindFork(lastFlushedNode)
+
+	// Disconnect all of the blocks back to the point of the fork.  This entails
+	// loading the blocks and their associated spent txos from the database and
+	// using that information to unspend all of the spent txos and remove the
+	// utxos created by the blocks.  In addition, if a block votes against its
+	// parent, the regular transactions are reconnected.
+	//
+	// Note that blocks will only need to be disconnected during initialization
+	// if an unclean shutdown occurred between a block being disconnected and the
+	// cache being flushed.  Since the cache is always flushed immediately after
+	// disconnecting a block, this will occur very infrequently.  In the typical
+	// catchup case, the fork node will be the last flushed node itself and this
+	// loop will be skipped.
+	view := NewUtxoViewpoint(b.utxoCache)
+	view.SetBestHash(&tip.hash)
+	var nextBlockToDetach *dcrutil.Block
+	n := lastFlushedNode
+	for n != nil && n != fork {
+		select {
+		case <-b.interrupt:
+			return errInterruptRequested
+		default:
+		}
+
+		// Grab the block to detach based on the node.  Use the fact that the
+		// blocks are being detached in reverse order, so the parent of the
+		// current block being detached is the next one being detached.
+		block := nextBlockToDetach
+		if block == nil {
+			var err error
+			block, err = b.fetchBlockByNode(n)
+			if err != nil {
+				return err
+			}
+		}
+		if n.hash != *block.Hash() {
+			panicf("detach block node hash %v (height %v) does not match "+
+				"previous parent block hash %v", &n.hash, n.height,
+				block.Hash())
+		}
+
+		// Grab the parent of the current block and also save a reference to it
+		// as the next block to detach so it doesn't need to be loaded again on
+		// the next iteration.
+		parent, err := b.fetchBlockByNode(n.parent)
+		if err != nil {
+			return err
+		}
+		nextBlockToDetach = parent
+
+		// Determine if treasury agenda is active.
+		isTreasuryEnabled, err := b.isTreasuryAgendaActive(n.parent)
+		if err != nil {
+			return err
+		}
+
+		// Load all of the spent txos for the block from the spend journal.
+		var stxos []spentTxOut
+		err = b.db.View(func(dbTx database.Tx) error {
+			stxos, err = dbFetchSpendJournalEntry(dbTx, block, isTreasuryEnabled)
+			return err
+		})
```

Comment:
Maybe fine, but this is setting an error outside of the function's scope and returning. Logically confusing for me.

Reply:
Not sure what you mean.  `View` takes a callback function as an argument.  The error return is from the callback which `View` passes along to the caller.

Reply:
I think you mean that since `err` is already defined, the callback function passed to view is setting it directly but returning it anyway?  I don't really see this as confusing myself, but we could redefine `err` in the callback function so that it is scoped within it.  Unless others feel strongly about it, though, I'll leave it as-is since this is how it is handled all over the codebase.

Reply:
It's fine! Thanks for commenting.